### ISSUE TITLE
Delete S3 version doesnt seem to actually delete it

### ISSUE THREAD

**Issue by pmarsden** [rclone/rclone#7082]
#### What is the problem you are having with rclone?
Deleting an S3 version doesnt actually seem to delete anything

#### What is your rclone version (output from `rclone version`)
v1.62.2

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)
Linux, Centos 7.9

#### Which cloud storage system are you using? (e.g. Google Drive)
Wasabi [versioned bucket]

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

First I get the versions

`rclone lsl wasabi:pxtest/Backups/Computer/Excel/ --s3-versions --include Dummy-File*.xls`

    19456 2022-07-15 13:32:47.000000000 Dummy-File.xls
    19456 2023-06-18 23:57:55.000000000 Dummy-File-v2023-06-20-154700-000.xls

Then I try to delete the old version

`rclone delete wasabi:pxtest/Backups/Computer/Excel/Dummy-File-v2023-06-20-154700-000.xls -vv`

    2023/06/25 15:05:38 DEBUG : rclone: Version "v1.62.2" starting with parameters ["rclone" "delete" "wasabi:pxtest/Backups/Computer/Excel/Dummy-File-v2023-06-20-154700-000.xls" "-vv"]
    2023/06/25 15:05:38 DEBUG : Creating backend with remote "wasabi:pxtest/Backups/Computer/Excel/Dummy-File-v2023-06-20-154700-000.xls"
    2023/06/25 15:05:38 DEBUG : Using config file from "/root/.config/rclone/rclone.conf"
    2023/06/25 15:05:38 DEBUG : Resolving service "s3" region "us-east-1"
    2023/06/25 15:05:39 DEBUG : Waiting for deletions to finish
    2023/06/25 15:05:39 DEBUG : 4 go routines active

Then I get the versions again

`rclone lsl wasabi:pxtest/Backups/Computer/Excel/ --s3-versions --include Dummy-File*.xls`

    19456 2022-07-15 13:32:47.000000000 Dummy-File.xls
    19456 2023-06-18 23:57:55.000000000 Dummy-File-v2023-06-20-154700-000.xls

The file I attempted to delete still exists, the delete gave no failure messages, but didnt delete the old version.

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)
See above.

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by animosity22** [comment]
Check out the docs.

https://rclone.org/s3/

it's explained there.

**Comment by pmarsden** [comment]
Nope its not, at least not anywhere I can find in that HUGE page.
I did some text searches on it, and the only references I could find were to delete **all** old versions.
I'm not trying to do that, I'm trying to delete specific old versions, as I listed in my report.

If you think I'm missing something in those docs then perhaps you could give a better clue where to look
[ or better still what exact command I should run ].

**Comment by animosity22** [comment]
I meant to link the version part:

https://rclone.org/s3/#versions as my specific section got removed for some reason.

rclone delete wasabi:pxtest/Backups/Computer/Excel/Dummy-File-v2023-06-20-154700-000.xls -vv doesn't see the old version and in the output it doesn't delete anything.

The section right here is the relevant part to your question as you want to delete and version.

![image](https://github.com/rclone/rclone/assets/5088712/6f9b17b2-dac8-44a9-a261-40d39d992ef1)

**Comment by pmarsden** [comment]
Sorry, but that still doesnt appear to tell me how to delete **one specific** old version.

`rclone backend cleanup-hidden wasabi:pxtest/Backups/Computer/Excel/` 
Will delete **all** old versions under that path, not the specific "Dummy-File-v2023-06-20-154700-000.xls"

`rclone backend cleanup-hidden wasabi:pxtest/Backups/Computer/Excel/Dummy-File-v2023-06-20-154700-000.xls`
This does nothing at all - it doesnt delete the version, and doesnt report any errors.

I even tried this, but it also just tried to delete all the old versions in that path (at least thats what dry run reported).
`rclone backend cleanup-hidden wasabi:pxtest/Backups/Computer/Excel/ --include Dummy-File-v2023-06-20-154700-000.xls`

Also, as an asside, the standard delete should not just silently fail, it should at least tell me it cant delete the file (or even 'see' it).

I appreciate your help, but as far as I'm concerned, at least two ways I ought to be able to delete this old version file are broken/fail/non-functional and if there is some other way, its not at all clear or obvious in any docs.

**Comment by animosity22** [comment]
Documentation is done by people so if it's not obvious, that's because no one has submitted anything to correct it.

I can't see your computer so I can't see any output on what you are typing.

On the forums, we have plenty of S3 people but you posted a general use question here so you are missing some of the folks that generally would help out.

**Comment by pmarsden** [comment]
I did not post a question, I posted a bug/failure, "delete" does not delete a file as expected.
Since then I've found that "backend cleanup-hidden" does not delete single files as expected either.

For the record, the output of the backend command is this;
`rclone backend cleanup-hidden wasabi:pxtest/Backups/Computer/Excel/Dummy-File-v2022-01-29-154700-000.xls -vv`

    2023/06/26 10:57:33 DEBUG : rclone: Version "v1.62.2" starting with parameters ["rclone" "backend" "cleanup-hidden" "wasabi:pxtest/Backups/Computer/Excel/Dummy-File-v2022-01-29-154700-000.xls" "-vv"]
    2023/06/26 10:57:33 DEBUG : Using config file from "/root/.config/rclone/rclone.conf"
    2023/06/26 10:57:33 DEBUG : Resolving service "s3" region "us-east-1"
    2023/06/26 10:57:34 DEBUG : S3 bucket pxtest path Backups/Computer/Excel/Dummy-File-v2022-01-29-154700-000.xls: bucket is versioned: true
    2023/06/26 10:57:34 DEBUG : Waiting for deletions to finish
    2023/06/26 10:57:34 DEBUG : 4 go routines active

I'm away for a few days so I wont be able to reply again (if necesary) until next weekend.

**Comment by ncw** [comment]
> Then I try to delete the old version
> 
> `rclone delete wasabi:pxtest/Backups/Computer/Excel/Dummy-File-v2023-06-20-154700-000.xls -vv`

You'll need to use `--s3-versions` on this command line.

    rclone delete --s3-versions wasabi:pxtest/Backups/Computer/Excel/Dummy-File-v2023-06-20-154700-000.xls

As to why that didn't give an error? That is because `rclone delete` can delete a directory full of files too and it assumed the argument was a directory since it didn't find a file without `--s3-versions`.

You can use `rclone deletefile` instead and that will give an error.

**Comment by pmarsden** [comment]
I managed to test this while away, it still does not work :(

`rclone delete --s3-versions wasabi:pxtest/Backups/Computer/Excel/Dummy-File-v2022-01-29-154700-000.xls -vv`

    2023/06/28 14:31:34 DEBUG : rclone: Version "v1.62.2" starting with parameters ["rclone" "delete" "--s3-versions" "wasabi:pxtest/Backups/Computer/Excel/Dummy-File-v2022-01-29-154700-000.xls" "-vv"]
    2023/06/28 14:31:34 DEBUG : Creating backend with remote "wasabi:pxtest/Backups/Computer/Excel/Dummy-File-v2022-01-29-154700-000.xls"
    2023/06/28 14:31:34 DEBUG : Using config file from "/root/.config/rclone/rclone.conf"
    2023/06/28 14:31:34 DEBUG : wasabi: detected overridden config - adding "{pO73u}" suffix to name
    2023/06/28 14:31:34 DEBUG : Resolving service "s3" region "us-east-1"
    2023/06/28 14:31:35 DEBUG : fs cache: renaming cache item "wasabi:pxtest/Backups/Computer/Excel/Dummy-File-v2022-01-29-154700-000.xls" to be canonical "wasabi{pO73u}:pxtest/Backups/Computer/Excel/Dummy-File-v2022-01-29-154700-000.xls"
    2023/06/28 14:31:35 DEBUG : Waiting for deletions to finish
    2023/06/28 14:31:35 DEBUG : 4 go routines active

I dont know what the extra config stuff is - it appears to be added by the "--s3-versions" (not added if I remove that).

The file itself remains, undeleted.

**Comment by ncw** [comment]
This works fine when I try it with rclone v1.62.2

```
$ rclone lsf --s3-versions s3:rclone-versions
$ echo hello | rclone rcat s3:rclone-versions/hello.txt
$ echo hello1 | rclone rcat s3:rclone-versions/hello.txt
$ echo hello2 | rclone rcat s3:rclone-versions/hello.txt
$ rclone lsf --s3-versions s3:rclone-versions
hello-v2023-06-28-163746-000.txt
hello-v2023-06-28-163748-000.txt
hello.txt
$ rclone delete --s3-versions s3:rclone-versions/hello-v2023-06-28-163746-000.txt
$ rclone lsf --s3-versions s3:rclone-versions
hello-v2023-06-28-163748-000.txt
hello.txt
```

Can you do a similar test?

**Comment by pmarsden** [comment]
Sure, slightly different bucket name as I got rid of pxtest.

```
[root@vanadium ~]# rclone lsf --s3-versions wasabi:pxm/Test
[root@vanadium ~]# echo hello1 | rclone rcat wasabi:pxm/Test/hello.txt
[root@vanadium ~]# echo hello2 | rclone rcat wasabi:pxm/Test/hello.txt
[root@vanadium ~]# echo hello3 | rclone rcat wasabi:pxm/Test/hello.txt
[root@vanadium ~]# rclone lsf --s3-versions wasabi:pxm/Test
hello-v2023-06-30-143336-000.txt
hello-v2023-06-30-143343-000.txt
hello.txt
[root@vanadium ~]# rclone delete --s3-versions wasabi:pxm/Test/hello-v2023-06-30-143343-000.txt
[root@vanadium ~]# rclone lsf --s3-versions wasabi:pxm/Test
hello-v2023-06-30-143336-000.txt
hello-v2023-06-30-143343-000.txt
hello.txt
```

As you can see "hello-v2023-06-30-143343-000.txt" was not deleted.

This is with rclone "v1.62.2", running on "centos 7.9.2009 (64 bit)"

**Comment by ncw** [comment]
I tried exactly your test on AWS S3 and it worked fine. Which is making me think this is a wasabi specific problem.

However I tried the test on wasabi and it worked fine too.

Can you try your test on AWS?

Can you try the recently released v1.63.0?

Can you do this part `rclone delete --s3-versions wasabi:pxm/Test/hello-v2023-06-30-143343-000.txt` with the flags `-vv --dump bodies` and post the output please?

Thanks

**Comment by pmarsden** [comment]
I dont have any buckets on AWS (versioned or otherwise).

I was not aware there was a 1.63.0 - I can certainly give that a try (with the mentioned flags, if it fails).

**Comment by pmarsden** [comment]
I have tested this with 1.63.0 and it now appears to work as expected.

I guess this is related to "Fix --s3-versions on individual objects (Nick Craig-Wood)" in the release notes ?

**Comment by ncw** [comment]
Yes 542677d8074fcac91f1caffdaf9d5576da4012b6 looks like the fix and it also explains why my tests with v1.62.0 passed.

I'll close this now.
